### DISCUSSION TITLE
Edges with no source for circuit diagrams

### DISCUSSION THREAD

**Discussion by bjourne** [terrastruct/d2#1990] [Q&A]
Hi all. MWE:
```
box: "Full Adder"
inp: "" {
    style.stroke-width: 0
    style.fill: white
    width: 0
    height: 0
}
out: "" {
    style.stroke-width: 0
    style.fill: white
    width: 0
    height: 0
}
inp -> box: "carry in"
out: ""
box -> out: "carry out"
```
As you can see, I'm trying to create a diagram for a full adder. For that I need edges that are "external" and have no source vertex associated with them. Or a source vertex that is just a dot and not a box. Is something like this possible to accomplish in D2?

**Comment by alixander** [comment]
@bjourne you can just set it to a small circle. Does that work for your use case?

https://play.d2lang.com/?script=TM0xysMwDAXgXad4eP9_2tVDoUvv4diiFog42A5NKbl7sbx006cn8aKG1rh5fAhIpc8BaDls7BGlRmXbaFhYPZwzvST17HG9mDLLM3cjASedREs5PNxjV8U9Ja6OZN3-rc6PIip7_6WsG_5umG8x1PqGrG5cWedSjhFPzrjs3dE3AAD__w%3D%3D&

**Comment by bjourne** [comment]
Thanks, but honestly it doesn't look that great.

**Reply by alixander** [reply]
Can you draw out what you're looking for?

**Reply by stereobooster** [reply]
https://play.d2lang.com/?script=TM7BCsMgDAbge57ix_vKdvUw2GXvYVVmIFTRlLWMvvuoMtjNzz8hvxfXWmwWHwJC1vEAWnIlWniuXmL_ETdHsTCm681Bk8WtI0V-Jf2p6S5xysV51t3iSsBBB9GcNwvzXEXwCCFWQ7yUqRew52nKq_6Tl4LLHWPNu1p38GLOqd5iztsZD444r2roGwAA__8%3D&

```
style.opacity: 0
```

**Comment by bjourne** [comment]
I want to draw a diagram [like this](https://www.gatevidyalay.com/wp-content/uploads/2018/06/4-bit-Ripple-Carry-Adder.png). As you can see many of the arrows points to or from labels and not boxes.

**Reply by gwhitney** [reply]
Just use text nodes when you want the arrow to emanate from a "label". Here's just a bit of your example diagram: https://play.d2lang.com/?script=yswztFJwNObKzDPUK85ILEi1UihJrSjhyswzslJwAokboYgn5VdYKSi5lebkKDimpKQWxeS5KHHll5ZYKQQbg2g0QwwVdO0UkvIrQObAmGDVSsHGSiDDQIL5pSVcgAAAAP__&
